File: go/error-handling/02_constraint_violation.go
package main

import (
	"database/sql"
	"errors"
	"fmt"

	_ "github.com/cubrid-labs/cubrid-go"
	cubrid "github.com/cubrid-labs/cubrid-go"
)

var ErrDuplicateKey = errors.New("duplicate key")

func classifyConstraintError(err error) error {
	if err == nil {
		return nil
	}

	var integrityErr *cubrid.IntegrityError
	if errors.As(err, &integrityErr) {
		return fmt.Errorf("%w: code=%d message=%s", ErrDuplicateKey, integrityErr.Code, integrityErr.Message)
	}

	return err
}

func main() {
	db, err := sql.Open("cubrid", "cubrid://dba:@localhost:33000/testdb")
	if err != nil {
		panic(err)
	}
	defer db.Close()

	if _, err := db.Exec("DROP TABLE IF EXISTS cookbook_error_users"); err != nil {
		panic(err)
	}
	defer db.Exec("DROP TABLE IF EXISTS cookbook_error_users")

	if _, err := db.Exec(`
		CREATE TABLE cookbook_error_users (
			id INT AUTO_INCREMENT PRIMARY KEY,
			email VARCHAR(200) UNIQUE NOT NULL
		)
	`); err != nil {
		panic(err)
	}

	if _, err := db.Exec("INSERT INTO cookbook_error_users (email) VALUES (?)", "alice@example.com"); err != nil {
		panic(err)
	}

	_, err = db.Exec("INSERT INTO cookbook_error_users (email) VALUES (?)", "alice@example.com")
	classified := classifyConstraintError(err)
	if errors.Is(classified, ErrDuplicateKey) {
		fmt.Printf("duplicate key handled via errors.Is: %v\n", classified)
		return
	}

	panic("expected duplicate key error")
}

File: rust/error-handling/02_constraint_violation.rs
use cubrid_tokio::Client;

const DSN: &str = "cubrid://dba:@localhost:33000/testdb";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut client = Client::connect(DSN).await?;

    client
        .execute("DROP TABLE IF EXISTS cookbook_error_users", &[])
        .await?;
    client
        .execute(
            "CREATE TABLE cookbook_error_users (
                id INT AUTO_INCREMENT PRIMARY KEY,
                email VARCHAR(200) UNIQUE NOT NULL
            )",
            &[],
        )
        .await?;

    client
        .execute(
            "INSERT INTO cookbook_error_users (email) VALUES ('alice@example.com')",
            &[],
        )
        .await?;

    let duplicate = client
        .execute(
            "INSERT INTO cookbook_error_users (email) VALUES ('alice@example.com')",
            &[],
        )
        .await;

    match duplicate {
        Ok(_) => return Err("expected unique constraint violation".into()),
        Err(error) => {
            let message = error.to_string().to_lowercase();
            if message.contains("unique") || message.contains("constraint") {
                println!("Constraint violation handled: {error}");
            } else {
                return Err(error.into());
            }
        }
    }

    client
        .execute("DROP TABLE IF EXISTS cookbook_error_users", &[])
        .await?;

    Ok(())
}
